File: sdks/sandbox/python/src/opensandbox/services/sandbox.py
#
# Copyright 2025 Alibaba Group Holding Ltd.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
"""
Sandbox service interface.

Protocol for sandbox lifecycle management operations.
"""

from datetime import datetime, timedelta
from typing import Protocol

from opensandbox.models.sandboxes import (
    CreateSnapshotRequest,
    CredentialProxyConfig,
    NetworkPolicy,
    PagedSandboxInfos,
    PagedSnapshotInfos,
    PlatformSpec,
    SandboxCreateResponse,
    SandboxEndpoint,
    SandboxFilter,
    SandboxImageSpec,
    SandboxInfo,
    SandboxLifecycle,
    SandboxRenewResponse,
    SnapshotFilter,
    SnapshotInfo,
    Volume,
)
from opensandbox.models.templates import (
    CreateTemplateRequest,
    PagedTemplateInfos,
    TemplateFilter,
    TemplateInfo,
)


class Sandboxes(Protocol):
    """
    Core sandbox lifecycle management service.

    This service provides a clean abstraction over sandbox creation, management,
    and termination operations, completely isolating business logic from API implementation details.
    """

    async def create_sandbox(
        self,
        spec: SandboxImageSpec | None,
        entrypoint: list[str] | None,
        env: dict[str, str],
        metadata: dict[str, str],
        timeout: timedelta | None,
        resource: dict[str, str],
        network_policy: NetworkPolicy | None,
        extensions: dict[str, str],
        volumes: list[Volume] | None,
        platform: PlatformSpec | None = None,
        secure_access: bool = False,
        snapshot_id: str | None = None,
        credential_proxy: CredentialProxyConfig | None = None,
        resource_requests: dict[str, str] | None = None,
        lifecycle: SandboxLifecycle | None = None,
    ) -> SandboxCreateResponse:
        """
        Create a new sandbox with the specified configuration.

        Args:
            spec: Container image specification for provisioning the sandbox.
            entrypoint: Command to run as the sandbox's main process.
            env: Environment variables injected into the sandbox runtime.
            metadata: User-defined metadata used for management and filtering.
            timeout: Sandbox lifetime. Pass None to create a sandbox that requires explicit cleanup.
            resource: Runtime resource limits (e.g. cpu/memory). Exact semantics are server-defined.
            network_policy: Optional outbound network policy (egress).
            credential_proxy: Optional Credential Vault proxy startup settings.
            extensions: Opaque extension parameters passed through to the server as-is.
                Prefer namespaced keys (e.g. ``storage.id``).
            volumes: Optional list of volume mounts for persistent storage.
            secure_access: Whether to enable secured access for sandbox endpoints.
            lifecycle: Optional pre-start and periodic lifecycle hooks.

        Returns:
            Sandbox create response

        Raises:
            SandboxException: if the operation fails
        """
        ...

    async def create_sandbox_from_template(
        self,
        template_id: str,
        timeout: timedelta,
        metadata: dict[str, str] | None = None,
        network_policy: NetworkPolicy | None = None,
        extensions: dict[str, str] | None = None,
    ) -> SandboxCreateResponse:
        """
        Create a sandbox from a ``Succeeded`` fsb template.

        Template mode fixes the workload shape on the server: only metadata,
        network policy and extensions may accompany the template id, and the
        timeout is required.

        Args:
            template_id: Succeeded fsb template to create the sandbox from
            timeout: Sandbox lifetime (required in template mode)
            metadata: User-defined metadata used for management and filtering
            network_policy: Optional outbound network policy (egress)
            extensions: Opaque extension parameters passed through to the server as-is

        Returns:
            Sandbox create response

        Raises:
            SandboxException: if the operation fails
        """
        ...

    async def get_sandbox_info(self, sandbox_id: str) -> SandboxInfo:
        """
        Retrieve information about an existing sandbox.

        Args:
            sandbox_id: Unique identifier of the sandbox

        Returns:
            Current sandbox information

        Raises:
            SandboxException: if the operation fails
        """
        ...

    async def list_sandboxes(self, filter: SandboxFilter) -> PagedSandboxInfos:
        """
        List sandboxes with optional filtering.

        Args:
            filter: Optional filter criteria

        Returns:
            List of sandbox information matching the filter

        Raises:
            SandboxException: if the operation fails
        """
        ...

    async def patch_sandbox_metadata(
        self, sandbox_id: str, patch: dict[str, str | None]
    ) -> SandboxInfo:
        """
        Patch sandbox metadata.

        Args:
            sandbox_id: Unique identifier of the sandbox
            patch: Metadata merge patch. String values add or replace keys; None deletes keys.

        Returns:
            Current sandbox information after applying the patch

        Raises:
            SandboxException: if the operation fails
        """
        ...

    async def get_sandbox_endpoint(
        self, sandbox_id: str, port: int, use_server_proxy: bool = False
    ) -> SandboxEndpoint:
        """
        Get sandbox endpoint.

        Args:
            sandbox_id: Sandbox ID
            port: Endpoint port number
            use_server_proxy: Whether to use server proxy for endpoint

        Returns:
            Target sandbox endpoint

        Raises:
            SandboxException: if the operation fails
        """
        ...

    async def get_signed_sandbox_endpoint(
        self, sandbox_id: str, port: int, expires: int,
        use_server_proxy: bool = False,
    ) -> SandboxEndpoint:
        """
        Get signed sandbox endpoint with an OSEP-0011 route token.

        Args:
            sandbox_id: Sandbox ID
            port: Endpoint port number
            expires: Unix epoch seconds for the signed route token expiry
            use_server_proxy: Whether to use server proxy for endpoint

        Returns:
            Target sandbox endpoint

        Raises:
            SandboxException: if the operation fails
        """
        ...

    async def pause_sandbox(self, sandbox_id: str) -> None:
        """
        Pause a running sandbox, preserving its state.

        Args:
            sandbox_id: Unique identifier of the sandbox

        Raises:
            SandboxException: if the operation fails
        """
        ...

    async def resume_sandbox(self, sandbox_id: str) -> None:
        """
        Resume a paused sandbox.

        Args:
            sandbox_id: Unique identifier of the sandbox

        Raises:
            SandboxException: if the operation fails
        """
        ...

    async def renew_sandbox_expiration(
        self, sandbox_id: str, new_expiration_time: datetime
    ) -> SandboxRenewResponse:
        """
        Renew the expiration time of a sandbox.

        Args:
            sandbox_id: Unique identifier of the sandbox
            new_expiration_time: New expiration timestamp

        Returns:
            Renew response including the new expiration time.

        Raises:
            SandboxException: if the operation fails
        """
        ...

    async def kill_sandbox(self, sandbox_id: str) -> None:
        """
        Terminate a sandbox and release all associated resources.

        Args:
            sandbox_id: Unique identifier of the sandbox

        Raises:
            SandboxException: if the operation fails
        """
        ...

    async def create_snapshot(
        self, sandbox_id: str, request: CreateSnapshotRequest | None = None
    ) -> SnapshotInfo:
        """Create a persistent snapshot from a sandbox."""
        ...

    async def get_snapshot(self, snapshot_id: str) -> SnapshotInfo:
        """Retrieve information about an existing snapshot."""
        ...

    async def list_snapshots(self, filter: SnapshotFilter) -> PagedSnapshotInfos:
        """List snapshots with optional filtering."""
        ...

    async def delete_snapshot(self, snapshot_id: str) -> None:
        """Delete a snapshot."""
        ...

    async def create_template(self, request: CreateTemplateRequest) -> TemplateInfo:
        """
        Create a fsb template (golden-image build).

        The build runs asynchronously: the response carries
        ``status.phase: Pending``; poll ``get_template`` until ``Succeeded``
        (or ``Failed``). Only ``Succeeded`` templates can back template-based
        sandbox creation.

        Raises:
            SandboxException: if the operation fails
        """
        ...

    async def get_template(self, template_id: str) -> TemplateInfo:
        """
        Get one template with its latest build status.

        Raises:
            SandboxException: if the operation fails
        """
        ...

    async def list_templates(self, filter: TemplateFilter) -> PagedTemplateInfos:
        """
        List the current tenant's templates with optional metadata filtering
        (AND logic) and pagination.

        Raises:
            SandboxException: if the operation fails
        """
        ...

    async def delete_template(self, template_id: str) -> None:
        """
        Delete a template.

        Sandboxes already created from the template are unaffected.

        Raises:
            SandboxException: if the operation fails
        """
        ...

    def invalidate_endpoint_cache(self, sandbox_id: str) -> None:
        """Remove all cached endpoints for a sandbox. No-op if caching is disabled."""
        ...

File: sdks/sandbox/go/sandbox.go
// Copyright 2026 Alibaba Group Holding Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package opensandbox

import (
	"context"
	"errors"
	"fmt"
	"strings"
	"sync"
	"time"
)

// SandboxCreateOptions configures sandbox creation.
type SandboxCreateOptions struct {
	// Image is the container image URI (required).
	Image string
	// SnapshotID restores the sandbox from a previously created snapshot.
	SnapshotID string

	// Entrypoint is the command to run. Defaults to DefaultEntrypoint.
	Entrypoint []string

	// Resource limits (e.g. {"cpu": "500m", "memory": "256Mi"}).
	// Defaults to DefaultResourceLimits.
	ResourceLimits ResourceLimits

	// ResourceRequests sets Kubernetes resource requests (guaranteed minimums).
	// When set, enables Burstable QoS (requests < limits).
	// When nil, limits are used for both limits and requests (Guaranteed QoS).
	ResourceRequests ResourceLimits

	// TimeoutSeconds is the sandbox TTL. Nil means use DefaultTimeoutSeconds.
	TimeoutSeconds *int

	// Env variables injected into the sandbox.
	Env map[string]string

	// SecureAccess enables secured access for sandbox endpoints.
	SecureAccess bool

	// Metadata for filtering and tagging.
	Metadata map[string]string

	// Lifecycle contains optional pre-start and periodic hooks.
	Lifecycle *SandboxLifecycle

	// NetworkPolicy for egress control.
	NetworkPolicy *NetworkPolicy

	// CredentialProxy enables Credential Vault transparent proxy support.
	CredentialProxy *CredentialProxyConfig

	// Volumes to mount.
	Volumes []Volume

	// ImageAuth provides registry credentials for private images.
	ImageAuth *ImageAuth

	// ManualCleanup, when true, creates the sandbox with no TTL so it stays
	// alive until explicitly killed. The timeout field is omitted from the
	// request (nil), causing the server to treat it as infinite.
	ManualCleanup bool

	// Extensions for provider-specific parameters.
	Extensions map[string]string

	// Platform selects the target OS/arch for the sandbox (e.g. {"os":
	// "windows", "arch": "amd64"}). When nil the server applies its default.
	Platform *PlatformSpec

	// SkipHealthCheck skips the WaitUntilReady call after creation.
	SkipHealthCheck bool

	// ReadyTimeout overrides DefaultReadyTimeoutSeconds.
	ReadyTimeout time.Duration

	// HealthCheckInterval overrides DefaultHealthCheckPollingInterval.
	HealthCheckInterval time.Duration

	// HealthCheck is a custom health check function. If nil, execd /ping is used.
	HealthCheck func(ctx context.Context, sb *Sandbox) (bool, error)
}

// Sandbox is the high-level object wrapping lifecycle, execd, and egress clients.
// Use CreateSandbox or ConnectSandbox to obtain an instance.
type Sandbox struct {
	id     string
	config *ConnectionConfig

	lifecycle *LifecycleClient
	execd     *ExecdClient
	egress    *EgressClient
	origin    SandboxOrigin
	mu        sync.Mutex
}

// ID returns the sandbox identifier.
func (s *Sandbox) ID() string { return s.id }

// Origin reports what backs this sandbox (see SandboxOrigin).
//
// SandboxOriginTemplate when the sandbox runs on a fsb golden-image template:
// set locally by CreateSandboxFromTemplate, and reported by the server's
// OPEN-SANDBOX-ORIGIN response header otherwise (also honored for snapshot
// restores, which boot the template's published artifact set).
// SandboxOriginUnknown for everything else.
//
// Template-backed sandboxes route egress policy operations through the
// lifecycle control plane (/sandboxes/{id}/networkpolicy) instead of the
// sandbox-side egress sidecar.
func (s *Sandbox) Origin() SandboxOrigin {
	if s.origin == "" {
		return SandboxOriginUnknown
	}
	return s.origin
}

// templateBacked reports whether this sandbox runs on a fsb golden-image
// template and therefore has no sandbox-side egress sidecar.
func (s *Sandbox) templateBacked() bool {
	return s.origin == SandboxOriginTemplate
}

// CreateSandbox creates a new sandbox and waits for it to be ready.
func CreateSandbox(ctx context.Context, config ConnectionConfig, opts SandboxCreateOptions) (*Sandbox, error) {
	if (opts.Image == "") == (opts.SnapshotID == "") {
		return nil, &InvalidArgumentError{Field: "Image/SnapshotID", Message: "exactly one of image or snapshotID is required"}
	}

	entrypoint := opts.Entrypoint
	if len(entrypoint) == 0 {
		entrypoint = DefaultEntrypoint
	}
	limits := opts.ResourceLimits
	if limits == nil {
		limits = DefaultResourceLimits
	}
	var timeout *int
	if opts.ManualCleanup {
		// nil timeout — omitted from JSON via omitempty, server treats as no TTL.
	} else if opts.TimeoutSeconds != nil {
		timeout = opts.TimeoutSeconds
	} else {
		t := DefaultTimeoutSeconds
		timeout = &t
	}

	lc := config.lifecycleClient()
	startupSource := opts.Image
	if startupSource == "" {
		startupSource = opts.SnapshotID
	}
	started := time.Now()

	req := CreateSandboxRequest{
		Image:            nil,
		SnapshotID:       opts.SnapshotID,
		Entrypoint:       entrypoint,
		ResourceLimits:   limits,
		ResourceRequests: opts.ResourceRequests,
		Timeout:          timeout,
		Env:              opts.Env,
		SecureAccess:     opts.SecureAccess,
		Metadata:         opts.Metadata,
		Lifecycle:        opts.Lifecycle,
		NetworkPolicy:    opts.NetworkPolicy,
		CredentialProxy:  opts.CredentialProxy,
		Volumes:          opts.Volumes,
		Extensions:       opts.Extensions,
		Platform:         opts.Platform,
	}
	if opts.Image != "" {
		req.Image = &ImageSpec{URI: opts.Image, Auth: opts.ImageAuth}
	}

	created, err := lc.CreateSandbox(ctx, req)
	if err != nil {
		reportSandboxCreateMetric(config, "", startupSource, time.Since(started).Milliseconds(), false)
		return nil, fmt.Errorf("opensandbox: create sandbox: %w", err)
	}

	return finishCreate(ctx, config, lc, created, startupSource, SandboxOriginUnknown, opts.readyOptions())
}

// SandboxFromTemplateOptions configures template-based sandbox creation.
type SandboxFromTemplateOptions struct {
	// TimeoutSeconds is the sandbox TTL. Required: the server rejects
	// template-based creation without a timeout.
	TimeoutSeconds int

	// Metadata for filtering and tagging.
	Metadata map[string]string

	// NetworkPolicy for egress control.
	NetworkPolicy *NetworkPolicy

	// Extensions for provider-specific parameters.
	Extensions map[string]string

	// SkipHealthCheck skips the WaitUntilReady call after creation.
	SkipHealthCheck bool

	// ReadyTimeout overrides DefaultReadyTimeoutSeconds.
	ReadyTimeout time.Duration

	// HealthCheckInterval overrides DefaultHealthCheckPollingInterval.
	HealthCheckInterval time.Duration

	// HealthCheck is a custom health check function. If nil, execd /ping is used.
	HealthCheck func(ctx context.Context, sb *Sandbox) (bool, error)
}

// CreateSandboxFromTemplate creates a new sandbox from a Succeeded fsb
// template and waits for it to be ready.
//
// Template mode fixes the workload shape on the server: the entrypoint, env,
// resources, volumes, platform and lifecycle of the sandbox come from the
// template's golden image and cannot be overridden here. Only metadata,
// network policy and extensions may accompany the template ID, and the
// timeout is required.
func CreateSandboxFromTemplate(ctx context.Context, config ConnectionConfig, templateID string, opts SandboxFromTemplateOptions) (*Sandbox, error) {
	if templateID == "" {
		return nil, &InvalidArgumentError{Field: "templateID", Message: "template ID is required"}
	}
	if opts.TimeoutSeconds <= 0 {
		return nil, &InvalidArgumentError{Field: "TimeoutSeconds", Message: "timeout is required when creating from a template"}
	}

	lc := config.lifecycleClient()
	startupSource := "template:" + templateID
	started := time.Now()

	req := CreateSandboxRequest{
		TemplateID:    templateID,
		Timeout:       &opts.TimeoutSeconds,
		Metadata:      opts.Metadata,
		NetworkPolicy: opts.NetworkPolicy,
		Extensions:    opts.Extensions,
	}

	created, err := lc.CreateSandbox(ctx, req)
	if err != nil {
		reportSandboxCreateMetric(config, "", startupSource, time.Since(started).Milliseconds(), false)
		return nil, fmt.Errorf("opensandbox: create sandbox from template: %w", err)
	}

	return finishCreate(ctx, config, lc, created, startupSource, SandboxOriginTemplate, opts.readyOptions())
}

// readyOptions is the shared subset of create options that controls the
// post-create readiness flow.
type readyOptions struct {
	skipHealthCheck     bool
	readyTimeout        time.Duration
	healthCheckInterval time.Duration
	healthCheck         func(ctx context.Context, sb *Sandbox) (bool, error)
}

func (o SandboxCreateOptions) readyOptions() readyOptions {
	return readyOptions{
		skipHealthCheck:     o.SkipHealthCheck,
		readyTimeout:        o.ReadyTimeout,
		healthCheckInterval: o.HealthCheckInterval,
		healthCheck:         o.HealthCheck,
	}
}

func (o SandboxFromTemplateOptions) readyOptions() readyOptions {
	return readyOptions{
		skipHealthCheck:     o.SkipHealthCheck,
		readyTimeout:        o.ReadyTimeout,
		healthCheckInterval: o.HealthCheckInterval,
		healthCheck:         o.HealthCheck,
	}
}

// finishCreate completes the shared create flow: wait for the sandbox to
// reach Running, resolve execd, verify readiness, and report create metrics.
// origin is the locally known sandbox origin (SandboxOriginUnknown unless the
// caller created the sandbox from a template); the server can refine it via
// the OPEN-SANDBOX-ORIGIN header during execd resolution. On failure the
// created sandbox is deleted best-effort.
func finishCreate(ctx context.Context, config ConnectionConfig, lc *LifecycleClient, created *SandboxInfo, startupSource string, origin SandboxOrigin, opts readyOptions) (*Sandbox, error) {
	started := time.Now()

	sb := &Sandbox{
		id:        created.ID,
		config:    &config,
		lifecycle: lc,
		origin:    origin,
	}

	if err := sb.waitForRunning(ctx, opts.readyTimeout); err != nil {
		// Best-effort cleanup
		_ = lc.DeleteSandbox(context.Background(), created.ID)
		reportSandboxCreateMetric(config, created.ID, startupSource, time.Since(started).Milliseconds(), false)
		return nil, err
	}

	if err := sb.resolveExecd(ctx); err != nil {
		_ = lc.DeleteSandbox(context.Background(), created.ID)
		reportSandboxCreateMetric(config, created.ID, startupSource, time.Since(started).Milliseconds(), false)
		return nil, fmt.Errorf("opensandbox: resolve execd: %w", err)
	}

	if !opts.skipHealthCheck {
		readyOpts := ReadyOptions{
			Timeout:         opts.readyTimeout,
			PollingInterval: opts.healthCheckInterval,
			HealthCheck:     opts.healthCheck,
		}
		if err := sb.WaitUntilReady(ctx, readyOpts); err != nil {
			_ = lc.DeleteSandbox(context.Background(), created.ID)
			reportSandboxCreateMetric(config, created.ID, startupSource, time.Since(started).Milliseconds(), false)
			return nil, err
		}
	}
	reportSandboxCreateMetric(config, created.ID, startupSource, time.Since(started).Milliseconds(), true)

	return sb, nil
}

// ConnectSandbox connects to an existing sandbox by ID.
func ConnectSandbox(ctx context.Context, config ConnectionConfig, sandboxID string, opts ...ReadyOptions) (*Sandbox, error) {
	if sandboxID == "" {
		return nil, &InvalidArgumentError{Field: "sandboxID", Message: "sandbox ID is required"}
	}
	if len(opts) > 1 {
		return nil, &InvalidArgumentError{
			Field:   "opts",
			Message: "at most one ReadyOptions is supported",
		}
	}

	lc := config.lifecycleClient()

	sb := &Sandbox{
		id:        sandboxID,
		config:    &config,
		lifecycle: lc,
	}

	ready := ReadyOptions{}
	if len(opts) > 0 {
		ready = opts[0]
	}
	timeout, interval := readinessDurations(ready)
	deadline := time.Now().Add(timeout)
	waitCtx, cancel := context.WithDeadline(ctx, deadline)
	defer cancel()
	var lastErr error
	for {
		if err := ctx.Err(); err != nil {
			return nil, err
		}
		if !time.Now().Before(deadline) {
			return nil, &SandboxReadyTimeoutError{SandboxID: sandboxID, Elapsed: timeout.String(), LastErr: lastErr}
		}
		err := sb.resolveExecd(waitCtx)
		if ctx.Err() != nil {
			return nil, ctx.Err()
		}
		if err == nil && time.Now().Before(deadline) {
			break
		}
		if err != nil && waitCtx.Err() != nil && lastErr == nil {
			lastErr = err
		}
		if err != nil && waitCtx.Err() == nil {
			var apiErr *APIError
			if !errors.As(err, &apiErr) || apiErr.StatusCode != 404 || apiErr.Response.Code != "KUBERNETES::POD_IP_NOT_AVAILABLE" {
				return nil, err
			}
			lastErr = err
		}
		select {
		case <-waitCtx.Done():
		case <-time.After(readinessSleep(interval, deadline)):
		}
	}
	if len(opts) > 0 {
		if err := sb.waitUntilReady(ctx, ready, deadline, timeout); err != nil {
			return nil, err
		}
	}

	return sb, nil
}

// ResumeSandbox resumes a paused sandbox and reconnects to it.
func ResumeSandbox(ctx context.Context, config ConnectionConfig, sandboxID string, opts ...ReadyOptions) (*Sandbox, error) {
	lc := config.lifecycleClient()
	if err := lc.ResumeSandbox(ctx, sandboxID); err != nil {
		return nil, fmt.Errorf("opensandbox: resume sandbox: %w", err)
	}
	return ConnectSandbox(ctx, config, sandboxID, opts...)
}

// Resume resumes this sandbox if it was paused and reconnects to it.
func (s *Sandbox) Resume(ctx context.Context, opts ...ReadyOptions) (*Sandbox, error) {
	return ResumeSandbox(ctx, *s.config, s.id, opts...)
}

// Kill terminates the sandbox. This is irreversible.
func (s *Sandbox) Kill(ctx context.Context) error {
	if s.lifecycle.cache != nil {
		s.lifecycle.cache.Invalidate(s.id)
	}
	return s.lifecycle.DeleteSandbox(ctx, s.id)
}

// Close is a no-op; it does not terminate the sandbox.
func (s *Sandbox) Close() error {
	return nil
}

// Pause pauses the sandbox while preserving its state.
// Endpoint cache is invalidated because endpoints may change across pause/resume.
func (s *Sandbox) Pause(ctx context.Context) error {
	if s.lifecycle.cache != nil {
		s.lifecycle.cache.Invalidate(s.id)
	}
	return s.lifecycle.PauseSandbox(ctx, s.id)
}

// GetInfo returns the sandbox's current info (status, metadata, image, etc.).
func (s *Sandbox) GetInfo(ctx context.Context) (*SandboxInfo, error) {
	return s.lifecycle.GetSandbox(ctx, s.id)
}

// PatchMetadata patches this sandbox's metadata. Non-nil values add or replace
// keys. Nil values delete keys.
func (s *Sandbox) PatchMetadata(ctx context.Context, patch MetadataPatch) (*SandboxInfo, error) {
	return s.lifecycle.PatchSandboxMetadata(ctx, s.id, patch)
}

// IsHealthy checks whether the sandbox's execd service is responsive.
func (s *Sandbox) IsHealthy(ctx context.Context) bool {
	if s.execd == nil {
		return false
	}
	return s.execd.Ping(ctx) == nil
}

// Ping checks if the execd service is responsive.
func (s *Sandbox) Ping(ctx context.Context) error {
	if s.execd == nil {
		return fmt.Errorf("opensandbox: execd client not initialized")
	}
	return s.execd.Ping(ctx)
}

// Renew extends the sandbox's expiration by the given duration from now.
func (s *Sandbox) Renew(ctx context.Context, duration time.Duration) (*RenewExpirationResponse, error) {
	return s.lifecycle.RenewExpiration(ctx, s.id, time.Now().Add(duration))
}

// CreateSnapshot creates a persistent snapshot from this sandbox.
func (s *Sandbox) CreateSnapshot(ctx context.Context, req CreateSnapshotRequest) (*SnapshotInfo, error) {
	return s.lifecycle.CreateSnapshot(ctx, s.id, req)
}

// GetEndpoint retrieves the public access endpoint for a service port.
func (s *Sandbox) GetEndpoint(ctx context.Context, port int) (*Endpoint, error) {
	useProxy := s.config.UseServerProxy
	return s.lifecycle.GetEndpoint(ctx, s.id, port, &useProxy)
}

// GetSignedEndpoint retrieves a signed endpoint URL with an OSEP-0011 route
// token that expires at the given Unix epoch timestamp (seconds).
func (s *Sandbox) GetSignedEndpoint(ctx context.Context, port int, expires int64) (*Endpoint, error) {
	return s.lifecycle.GetSignedEndpoint(ctx, s.id, port, expires)
}

// ReadyOptions configures WaitUntilReady behavior.
// Timely timeout requires custom health checks and transports to honor context cancellation.
type ReadyOptions struct {
	Timeout         time.Duration
	PollingInterval time.Duration
	HealthCheck     func(ctx context.Context, sb *Sandbox) (bool, error)
}

// WaitUntilReady polls until the sandbox is ready or the timeout expires.
// By default it checks execd /ping; if HealthCheck is provided, it uses that instead.
func (s *Sandbox) WaitUntilReady(ctx context.Context, opts ReadyOptions) error {
	timeout, _ := readinessDurations(opts)
	return s.waitUntilReady(ctx, opts, time.Now().Add(timeout), timeout)
}

func readinessDurations(opts ReadyOptions) (time.Duration, time.Duration) {
	timeout, interval := opts.Timeout, opts.PollingInterval
	if timeout == 0 {
		timeout = time.Duration(DefaultReadyTimeoutSeconds) * time.Second
	}
	if interval <= 0 {
		interval = DefaultHealthCheckPollingInterval
	}
	return timeout, interval
}

func (s *Sandbox) waitUntilReady(ctx context.Context, opts ReadyOptions, deadline time.Time, timeout time.Duration) error {
	_, interval := readinessDurations(opts)
	waitCtx, cancel := context.WithDeadline(ctx, deadline)
	defer cancel()
	var lastErr error
	for time.Now().Before(deadline) {
		if ctx.Err() != nil {
			return ctx.Err()
		}
		var healthy bool
		var err error
		if opts.HealthCheck != nil {
			healthy, err = opts.HealthCheck(waitCtx, s)
		} else {
			err = s.execd.Ping(waitCtx)
			healthy = err == nil
		}
		if ctx.Err() != nil {
			return ctx.Err()
		}
		if err != nil {
			lastErr = err
		}
		if healthy && time.Now().Before(deadline) {
			return nil
		}
		select {
		case <-waitCtx.Done():
		case <-time.After(readinessSleep(interval, deadline)):
		}
	}
	if ctx.Err() != nil {
		return ctx.Err()
	}
	return &SandboxReadyTimeoutError{SandboxID: s.id, Elapsed: timeout.String(), LastErr: lastErr}
}

func (s *Sandbox) waitForRunning(ctx context.Context, timeout time.Duration) error {
	if timeout <= 0 {
		timeout = time.Duration(DefaultReadyTimeoutSeconds) * time.Second
	}

	waitCtx := ctx
	cancel := func() {}
	if _, hasDeadline := ctx.Deadline(); !hasDeadline {
		waitCtx, cancel = context.WithTimeout(ctx, timeout)
	}
	defer cancel()

	start := time.Now()
	for {
		if err := waitCtx.Err(); err != nil {
			if errors.Is(err, context.DeadlineExceeded) {
				return &SandboxRunningTimeoutError{
					SandboxID: s.id,
					Elapsed:   time.Since(start).String(),
					LastErr:   err,
				}
			}
			return fmt.Errorf("opensandbox: sandbox %s did not reach Running state: %w", s.id, err)
		}

		info, err := s.lifecycle.GetSandbox(waitCtx, s.id)
		if err != nil {
			return fmt.Errorf("opensandbox: get sandbox status: %w", err)
		}
		if info.Status.State == StateRunning {
			return nil
		}
		if info.Status.State == StateFailed || info.Status.State == StateTerminated {
			return fmt.Errorf("opensandbox: sandbox %s entered terminal state: %s (%s)",
				s.id, info.Status.State, info.Status.Reason)
		}
		select {
		case <-waitCtx.Done():
		case <-time.After(2 * time.Second):
		}
	}
}

func (s *Sandbox) resolveExecd(ctx context.Context) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.execd != nil {
		return nil
	}

	useProxy := s.config.UseServerProxy
	endpoint, err := s.lifecycle.getEndpointFromServer(ctx, s.id, DefaultExecdPort, &useProxy)
	if err != nil {
		return err
	}

	// The server is authoritative about the runtime backing: for fsb-
	// prefixed sandboxes it reports "template" even when the create used a
	// snapshotId (a restore boots the template's published artifact set).
	// Never overwrite locally known origin with a missing header.
	if endpoint.Origin != "" {
		s.origin = endpoint.Origin
	}

	execdURL := s.config.RewriteEndpointURL(endpoint.Endpoint)
	if !strings.HasPrefix(execdURL, "http") {
		execdURL = s.config.GetProtocol() + "://" + execdURL
	}

	headers := make(map[string]string, len(endpoint.Headers)+1)
	for k, v := range endpoint.Headers {
		headers[k] = v
	}
	if s.config.UseServerProxy {
		if _, ok := headers[execdAuthHeader]; !ok {
			if apiKey := s.config.GetAPIKey(); apiKey != "" {
				headers[execdAuthHeader] = apiKey
			}
		}
	}

	s.execd = s.config.execdClient(execdURL, headers)
	return nil
}

func (s *Sandbox) resolveEgress(ctx context.Context) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.egress != nil {
		return nil
	}

	useProxy := s.config.UseServerProxy
	endpoint, err := s.lifecycle.GetEndpoint(ctx, s.id, DefaultEgressPort, &useProxy)
	if err != nil {
		return err
	}

	egressURL := s.config.RewriteEndpointURL(endpoint.Endpoint)
	if !strings.HasPrefix(egressURL, "http") {
		egressURL = s.config.GetProtocol() + "://" + egressURL
	}

	headers := make(map[string]string, len(endpoint.Headers)+1)
	for k, v := range endpoint.Headers {
		headers[k] = v
	}
	if s.config.UseServerProxy {
		if _, ok := headers[egressAuthHeader]; !ok {
			if apiKey := s.config.GetAPIKey(); apiKey != "" {
				headers[egressAuthHeader] = apiKey
			}
		}
	}

	s.egress = s.config.egressClient(egressURL, headers)
	return nil
}

func readinessSleep(interval time.Duration, deadline time.Time) time.Duration {
	if remaining := time.Until(deadline); remaining < interval {
		return remaining
	}
	return interval
}
